docs/proposals/12-oidc-webhook-authenticator.md

```diff
@@ -0,0 +1,305 @@
+
+To workaround the one OIDC per Kube APIServer limitation, a new `OIDC Webhook Authenticator` (OWA) could be implemented.
+
+## Goals
+
+- Dynamic registrations of OpenID Connect configurations.
+- Close as possible to the Kubernetes build-in OIDC Authenticator`.
+- Build as an optional extension and not required for functional Shoot or Gardener cluster.
+
+## Non-goals
+
+- [Dynamic Authorization](https://kubernetes.io/docs/reference/access-authn-authz/webhook/) is out of scope.
+
+## Proposal
+
+The Kube APIServer can use [Webhook Token Authentication](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) to send a [Bearer Tokens (id_token)](https://tools.ietf.org/html/rfc6750#section-2.1) to external webhook for validation:
+
+```json
+{
+  "apiVersion": "authentication.k8s.io/v1beta1",
+  "kind": "TokenReview",
+  "spec": {
+    "token": "(BEARERTOKEN)"
+  }
+}
+```
+
+Where upon verification, the remote webhook returns the identity of the user (if authentication succeeds):
+
+```json
+{
+  "apiVersion": "authentication.k8s.io/v1beta1",
+  "kind": "TokenReview",
+  "status": {
+    "authenticated": true,
+    "user": {
+      "username": "janedoe@example.com",
+      "uid": "42",
+      "groups": [
+        "developers",
+        "qa"
+      ],
+      "extra": {
+        "extrafield1": [
+          "extravalue1",
+          "extravalue2"
+        ]
+      }
+    }
+  }
+}
+```
+
+### Registration of new OpenIDConnect
+
+This new OWA can be configured with multiple OIDC providers and the entire flow can look like this:
+
+1. Admin adds a new `OpenIDConnect` resource (via CRD) to the cluster.
+
+    ```yaml
+    apiVersion: authentication.gardener.cloud/v1alpha1
+    kind: OpenIDConnect
+    metadata:
+      name: foo
+    spec:
+      issuerURL: https://foo.bar
+      clientID: some-client-id
+      usernameClaim: email
+      usernamePrefix: "test-"
+      groupsClaim: groups
+      groupsPrefix: "baz-"
+      supportedSigningAlgs:
+      - RS256
+      requiredClaims:
+        baz: bar
+      caBundle: LS0tLS1CRUdJTiBDRVJU...base64-encoded cert
+    ```
+
+1. OWA watches for changes on this resource and does [OIDC discovery](https://openid.net/specs/openid-connect-discovery-1_0.html), getting `issuer` and `jwks_uri`.
+1. OWA fetches the keys from `jwks_uri` and stores them in the status of `OpenIDConnect`:
+
+    ```yaml
+    apiVersion: authentication.gardener.cloud/v1alpha1
+    kind: OpenIDConnect
+    metadata:
+      name: foo
+    spec:
+      issuerURL: https://foo.bar
+      ...
+    status:
+      keys: f31deA9b... #the content of jwks_uri base64-encoded
+    ```
+
+1. OWA uses those keys, issuer, client_id and other settings to add OIDC authenticator to a list of already existing authenticators.
+
+![alt text](assets/registration.svg "Authentication with OIDC webhook")
```

Comment:
Please take note that the svg doesn't render when viewing in the rich diff editor - you have to go `··· -> View File`